File: encrypt.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# version: release
from __future__ import division

import argparse
import base64
import copy
import dbm
import glob
import hashlib
import math
import os
import queue
import random
import sys
import threading
import time


def group(lst, n):
    num = len(lst) % n
    zipped = zip(*[iter(lst)] * n)
    return zipped if not num else zipped + [lst[-num:], ]


def partition(lst, n):
    return group(lst, int(math.ceil(len(lst)/n)))


def rand(min, max):
    return random.randint(min, max)


def md5(s: str, raw_output=False):
    """Calculates the md5 hash of a given string"""
    res = hashlib.md5(s.encode())
    if raw_output:
        return res.digest()
    return res.hexdigest()


def substr_replace(subject, replace, start, length):
    if length is None:
        return subject[:start] + replace
    elif length < 0:
        return subject[:start] + replace + subject[length:]
    else:
        return subject[:start] + replace + subject[start+length:]


def base64_decode(data: str):
    return base64.b64decode(data.encode()).decode()


def base64_encode(data: str):
    return base64.b64encode(data.encode()).decode()


def strpos(haystack, needle, offset=0):
    """ Rewrite this to make it same with php when return is '-1', Or it will cause some error someway. """
    result = haystack.find(needle, offset)
    result = result if result > -1 else False
    return result


class FileSafeException(Exception):
    """docstring for FileSafeException"""

    msg = ''

    def __init__(self, msg):
        super(FileSafeException, self).__init__()
        self.msg = msg

    def __str__(self):
        return self.msg

    @classmethod
    def getMsg(self):
        return self.msg


class FileSafe():
    """Encrypt binaryfile class helper"""

    arrps = [[0, 20], [5, 30], [20, 10]]
    bFileNameEncode = True
    bFileEncrypt = True
    bRecurse = False
    bDebug = False
    count = 0
    sdir = ''
    subdir = []

    @classmethod
    def setWorkQueue(self, work_queue):
        self.work_queue = work_queue

    @classmethod
    def setFileNameEncode(self, encode=True):
        self.bFileNameEncode = encode

    @classmethod
    def setFileEncrypt(self, encrypt=True):
        self.bFileEncrypt = encrypt

    @classmethod
    def setRecurse(self, recurse=True):
        self.bRecurse = recurse

    @classmethod
    def setDebug(self, debug=True):
        self.bDebug = debug

    @classmethod
    def strDecrypt(self, key):
        value = db.get(key).decode()
        return base64_decode(value)

    @classmethod
    def strEncrypt(self, txt):
        key = md5(txt);
        value = base64_encode(txt)
        db[key] = value
        return key

    @classmethod
    def getFiles(self, sdir):
        fileStore = []
        if os.path.exists(sdir):
            if self.bRecurse:
                for root, dirs, files in os.walk(sdir, True):
                    root = root.replace('\\', '/').rstrip('/')
                    sep = '/'
                    for file in files:
                        if file.endswith('.db'):
                            continue
                        fileStore.append(root + sep + file)
                    for sdir in dirs:
                        self.subdir.append(root + sep + sdir)
            else:
                sdir = os.name == 'nt' \
                    and sdir.replace('/', os.sep).rstrip(os.sep) \
                    or sdir.replace('\\', os.sep).rstrip(os.sep)
                self.sdir = sdir
                # fileStore = [file for file in glob.glob(sdir + os.sep + '*') if os.path.isfile(file)]
                for file in glob.glob(sdir + os.sep + '*'):
                    if file.endswith('.db'):
                        continue
                    if os.path.isfile(file):
                        fileStore.append(file)
                    elif os.path.isdir(file):
                        self.subdir.append(file)
        return fileStore

    @classmethod
    def encrypts(self, sdir):
        files = self.getFiles(sdir)
        if files:
            self.encryptsByList(files)

    @classmethod
    def decrypts(self, sdir):
        self.setFileEncrypt(False)
        self.encrypts(sdir)

    @classmethod
    def encryptsByList(self, filelist):
        for file in filelist:
            try:
                self.encrypt(file)
            except FileSafeException as e:
                self.work_queue.put(str(e))

    @classmethod
    def encrypt(self, file):
        if os.path.isfile(file):
            self.fileEncrypt(file)
            if self.bFileNameEncode:
                self.fileNameEncode(file)

    @classmethod
    def fileEncrypt(self, file):
        """ Notice: let 'copy.deepcopy' do it, 'arrps = self.arrps' will always change 'self.arrps',
        because variables passed by reference in python. You can also do it like: 'arrps = self.arrps[:]'
        """
        arrps = copy.deepcopy(self.arrps)
        if self.bFileEncrypt is False:
            arrps.reverse()
            for ps in arrps:
                ps.reverse()

        info = os.stat(file)
        mtime = info.st_mtime
        atime = info.st_atime

        try:
            with open(file, 'rb+') as f:
                for pos in arrps:
                    pos1 = pos[0]
                    pos2 = pos[1]
                    f.seek(pos1)
                    char1 = f.read(1)
                    f.seek(pos2)
                    char2 = f.read(1)

                    f.seek(pos1)
                    f.write(char2)
                    f.seek(pos2)
                    f.write(char1)
        except Exception as e:
            msg = "EncryptError: " + str(e)
            raise FileSafeException(msg)

        os.utime(file, (atime, mtime))
        self.count += 1

        tag = self.bFileEncrypt and 'encrypt' or 'decrypt'
        msg = "Filename '" + file + "' has been " + tag
        if self.bDebug:
            self.work_queue.put(msg)

    @classmethod
    def subDirNameEncode(self):
        self.subdir.reverse()
        for sdir in self.subdir:
            self.fileNameEncode(sdir)

    @classmethod
    def fileNameEncode(self, file):
        """ 'os.path.basename' sometime it got wrong idea (specail char in win?) """
        try:
            # basename = os.path.basename(file)
            # dirname  = os.path.dirname(file)
            dirname = self.bRecurse and file[:file.rindex('/') + 1] or self.sdir + os.sep
            basename = file.replace(dirname, '')
            basename = self.bFileEncrypt and self.strEncrypt(basename) or self.strDecrypt(basename)
            oldname = file
            newname = os.path.join(dirname, basename)
            os.rename(oldname, newname)

            msg = "File '" + oldname + "' has been rename '" + newname + "'."
            if self.bDebug:
                self.work_queue.put(msg)
        except Exception as e:
            msg = "!EncodeError: " + str(e) + \
                ", Do you action decrypt and filename still be not fileNameEncode ?" + \
                "\n*You better with option '-x'*"
            raise FileSafeException(msg)


class ThreadRunner(threading.Thread):
    '''@summary: __init__ function def

    @param work_queue: queue for sending thread status
    @param showStatus: whether print thread end status
    @param callback: FileSafe class
    @param parameter: files to be handle via FileSafe class
    '''
    def __init__(self, work_queue, showStatus, callback, *parameter):
        super(ThreadRunner, self).__init__()    # Notice: must call supper __init__
        self.showStatus = showStatus
        self.callback = callback
        self.parameter = parameter
        self.work_queue = work_queue

    def run(self):
        self.callback(*self.parameter)
        if self.showStatus:
            msg = " *" + self.name.ljust(9) + " is ending at: " + time.ctime()
            self.work_queue.put(msg)


class Printer(threading.Thread):
    """@summary: The queue is for printing msg
    @param work_queue: queue
    """
    def __init__(self, work_queue):
        super(Printer, self).__init__()
        self.work_queue = work_queue

    def run(self):
        while True:
            msg = self.work_queue.get()
            if msg == 'ENDTAG':
                break
            print(msg)

""" ################################################################ """
""" ##################### Main code ################################ """
""" ################################################################ """
if __name__ == "__main__":

    parser = argparse.ArgumentParser(description='Encrypt Or Decrypt binaryfile.')
    parser.add_argument('-e', action='store_true', default=True, help='encrypt binaryfile, default: True')
    parser.add_argument('-d', action='store_true', help='decrypt binaryfile')
    parser.add_argument('-x', action='store_true', default=False, help='dont encrypt filename')
    parser.add_argument('-r', action='store_true', default=False, help='recurse the dir')
    parser.add_argument('-s', action='store_true', help='run with single thread')
    parser.add_argument('-m', action='store_true', default=True, help='run with multiple thread, default: True')
    parser.add_argument('-a', action='store_true', default=False, help='subdir name encode, default: False')
    parser.add_argument('-p', action='store_true', default=False, help='run with debug mode and print msg?')
    parser.add_argument('-n', metavar='ThreadCount', type=int, default=50, help='thread count, default is 50')
    parser.add_argument('dir', metavar='Directory', nargs='?', help='the directory to be handle')

    args = parser.parse_args()

    # print args
    # sys.exit()

    if args.dir is None:
        parser.print_help()
        sys.exit()

    if os.path.exists(args.dir) is False:
        print("Please check the sdir, it may be not exists.")
        sys.exit()

    """ ########### >> init variables ################################### """
    work_queue = queue.Queue()
    tcount = args.n     # Max Thread Count
    sucessed = '''
     _____________
     < well done >
     -------------
        \\   ^__^
         \\  (oo)\\_______
            (__)\\       )\\/\\

                ||----w |
                ||     ||
    '''
    """ ########### << init variables ################################## """

    FileSafe.setWorkQueue(work_queue)
    FileSafe.setDebug(args.p)
    FileSafe.setFileNameEncode(not args.x)
    FileSafe.setRecurse(args.r)

    """ ################################################################ """
    """ #################### single thread main run #################### """
    """ ################################################################ """
    if args.s:
        start = time.time()

        if args.d:
            FileSafe.decrypts(args.dir)
        else:
            FileSafe.encrypts(args.dir)

        # finally subdir name handle
        if args.a:
            FileSafe.subDirNameEncode()

        end = time.time()
        print(sucessed)
        print("**Take times: ", 'seconds: %.3f' % (end - start))
        sys.exit()

    """ ################################################################ """
    """ #################### multiple thread main run ################## """
    """ ################################################################ """
    if args.d:
        FileSafe.setFileEncrypt(False)
    else:
        FileSafe.setFileEncrypt(True)

    files = FileSafe.getFiles(args.dir)
    if len(files) == 0:
        print('No files to be handle...')
        sys.exit()
    fpartition = partition(files, tcount)

    # open db
    dbfile = os.path.join(args.dir, 'data')
    db = dbm.open(dbfile,'c')
    
    work_queue.put("**Threading Starting ... %s ... " % time.ctime())
    start = time.time()

    # Thread start
    threads = []
    for p in fpartition:
        t = ThreadRunner(work_queue, True, FileSafe.encryptsByList, p)
        threads.append(t)
        t.daemon = True
        t.start()

    # Thread Printer
    printer = Printer(work_queue)
    printer.daemon = True  # make sure sub thread exit when main thread is die
    printer.start()

    # Thread end
    for t in threads:
        t.join()

    work_queue.put(sucessed)
    end = time.time()
    msg = "**Threading(%d) all ending at:%s, Take times:%.3f seconds" % (len(threads), time.ctime(), (end - start))
    work_queue.put(msg)

    # All to be end.
    work_queue.put("ENDTAG")
    printer.join()

    # finally subdir name handle
    if args.a:
            FileSafe.subDirNameEncode()
    
    # close db
    db.close()

    # remove db file after decrypt 
    if args.d:
        os.remove(dbfile + '.db')

    # work_queue.join()

File: encrypt.go
// Tips: go run *.go will print exit status, go build *.go and run with binary file will ignore it.
// vim: fdm=indent ts=4 sw=4 sts=4 expandtab
package main

import (
	"crypto/md5"
	b64 "encoding/base64"
	"encoding/hex"
	"fmt"
	"math/rand"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	s "strings"
	"sync"
	"syscall"
	"time"
)

var P = fmt.Println

// => comman func {{{

const (
	IKEY  string = "-x6g6ZWm2G9g_vr0Bo.pOq3kRIxsZ6rm"
	KEY   string = "!@#$%^&*"
	CHARS string = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_~"
)

func GetOptVal(opt string) string {
	i := ArraySearch(os.Args, opt)
	if i == -1 || len(os.Args) == i+1 {
		return ""
	}
	val := os.Args[i+1]
	if val[:1] == "-" && len(val) == 2 {
		return ""
	}
	return val
}

func ArraySearch(vs []string, t string) int {
	for i, v := range vs {
		if v == t {
			return i
		}
	}
	return -1
}

func GetPathFromArgs(vs []string) string {
	path := ""
	for _, v := range vs {
		if ExistsPath(v) {
			return v
		}
	}
	return path
}

func ExistsPath(name string) bool {
	if _, err := os.Stat(name); err != nil {
		if os.IsNotExist(err) {
			return false
		}
	}
	return true
}

func GetStatTimes(name string) (atime, mtime time.Time, err error) {
	fi, err := os.Stat(name)
	if err != nil {
		return
	}
	mtime = fi.ModTime()
	stat := fi.Sys().(*syscall.Stat_t)
	atime = time.Unix(int64(stat.Atimespec.Sec), int64(stat.Atimespec.Nsec))
	return
}

func Base64Encode(text string) string {
	sEnc := b64.StdEncoding.EncodeToString([]byte(text))
	return sEnc
}

func Base64Decode(text string) string {
	sDec, _ := b64.StdEncoding.DecodeString(text)
	return string(sDec)
}

func GetMD5Hash(text string) string {
	hasher := md5.New()
	hasher.Write([]byte(text))
	return hex.EncodeToString(hasher.Sum(nil))
}

func SubsrReplace(subject string, replace string, start int, leng ...int) string {
	str := ""
	if len(leng) == 0 {
		str = subject[:start] + replace
	} else {
		length := leng[0]
		if length < 0 {
			str = subject[:start] + replace + subject[length:]
		} else {
			str = subject[:start] + replace + subject[start+length:]
		}
	}
	return str
}

func Decode(txt string) string {
	key := KEY
	chars := CHARS
	ikey := IKEY
	knum := 0
	tlen := len(txt)
	clen := len(CHARS)

	for _, c := range key {
		knum += int(c)
	}

	ch1 := string(txt[knum%tlen])
	nh1 := s.Index(chars, ch1)
	txt = SubsrReplace(txt, "", knum%tlen, 1)
	tlen -= 1
	ch2 := string(txt[nh1%tlen])
	nh2 := s.Index(chars, ch2)
	txt = SubsrReplace(txt, "", nh1%tlen, 1)
	tlen -= 1
	ch3 := string(txt[nh2%tlen])
	nh3 := s.Index(chars, ch3)
	txt = SubsrReplace(txt, "", nh2%tlen, 1)
	nhnum := nh1 + nh2 + nh3
	mdKey := GetMD5Hash(GetMD5Hash(GetMD5Hash(key+ch1)+ch2+ikey) + ch3)[nhnum%8 : nhnum%8+knum%8+16]
	tmp := ""
	k := 0
	klen := len(mdKey)

	for _, char := range txt {
		if k == klen {
			k = 0
		}
		ci := s.Index(chars, string(char))
		if ci < 0 {
			ci = 0
		}
		j := ci - nhnum - int(mdKey[k])
		k += 1
		for j < 0 {
			j += 64
		}
		if j < 0 {
			tmp += string(chars[j+clen])
		} else {
			tmp += string(chars[j])
		}
	}
	keys := []string{"-", "_", "."}
	values := []string{"+", "/", "="}
	for i, char := range keys {
		tmp = s.Replace(tmp, char, values[i], -1)
	}
	tmp = Base64Decode(tmp)
	r := regexp.MustCompile("[\000]")
	tmp = r.ReplaceAllString(tmp, "")
	return tmp
}

func Encode(txt string) string {
	key := KEY
	chars := CHARS
	ikey := IKEY

	nh1 := rand.Intn(64)
	nh2 := rand.Intn(64)
	nh3 := rand.Intn(64)

	ch1 := string(chars[nh1])
	ch2 := string(chars[nh2])
	ch3 := string(chars[nh3])
	nhnum := nh1 + nh2 + nh3
	knum := 0
	for _, c := range key {
		knum += int(c)
	}
	mdKey := GetMD5Hash(GetMD5Hash(GetMD5Hash(key+ch1)+ch2+ikey) + ch3)[nhnum%8 : nhnum%8+knum%8+16]
	txt = Base64Encode(txt)

	keys := []string{"-", "_", "."}
	values := []string{"+", "/", "="}
	for i, char := range keys {
		txt = s.Replace(txt, char, values[i], -1)
	}

	tmp := ""
	k := 0
	j := 0
	klen := len(mdKey)
	clen := len(CHARS)

	for _, char := range txt {
		if k == klen {
			k = 0
		}
		ci := s.Index(chars, string(char))
		if ci < 0 {
			ci = 0
		}
		j = (nhnum + ci + int(mdKey[k])) % 64
		k += 1
		if j < 0 {
			tmp += string(chars[j+clen])
		} else {
			tmp += string(chars[j])
		}
	}

	tmplen := len(tmp)
	tmplen += 1
	tmp = SubsrReplace(tmp, ch3, nh2%tmplen, 0)
	tmplen += 1
	tmp = SubsrReplace(tmp, ch2, nh1%tmplen, 0)
	tmplen += 1
	tmp = SubsrReplace(tmp, ch1, knum%tmplen, 0)
	return tmp
}

func EncodeFilename(filename string) string {
	basename := filepath.Base(filename)
	dirname := filepath.Dir(filename)
	basename = Encode(basename)
	filename = filepath.Join(dirname, basename)
	return filename
}

func DecodeFilename(filename string) string {
	basename := filepath.Base(filename)
	dirname := filepath.Dir(filename)
	basename = Decode(basename)
	filename = filepath.Join(dirname, basename)
	return filename
}

func Encrypt(file string, action string, encfilename bool) error {
	arrps := [3][2]int{{0, 20}, {5, 30}, {20, 10}}
	if action == "decrypt" {
		arrps = [3][2]int{{10, 20}, {30, 5}, {20, 0}}
	}

	atime, mtime, err := GetStatTimes(file)
	if err != nil {
		P(err)
	}

	f, err := os.OpenFile(file, os.O_RDWR, 0755)
	if err != nil {
		P(err)
	}

	for _, pos := range arrps {
		pos1 := pos[0]
		pos2 := pos[1]
		f.Seek(int64(pos1), 0)
		char1 := make([]byte, 1)
		f.Read(char1)
		f.Seek(int64(pos2), 0)
		char2 := make([]byte, 1)
		f.Read(char2)

		f.Seek(int64(pos1), 0)
		f.Write(char2)
		f.Seek(int64(pos2), 0)
		f.Write(char1)

	}

	if err := f.Close(); err != nil {
		P(err)
	}

	if err := os.Chtimes(file, atime, mtime); err != nil {
		P(err)
	}

	if encfilename {
		nfile := ""
		if action == "decrypt" {
			nfile = DecodeFilename(file)
		} else {
			nfile = EncodeFilename(file)
		}
		os.Rename(file, nfile)
	}

	return nil
}

func GetFilelist(path string) []string {
	filelist := []string{}
	filepath.Walk(path, func(path string, f os.FileInfo, err error) error {
		if !f.IsDir() {
			filelist = append(filelist, path)
		}
		return nil
	})
	return filelist
}

// => common func block end }}}

// => pool block {{{
type Pool struct {
	queue chan int
	wg    *sync.WaitGroup
}

func NewPool(cap, total int) *Pool {
	if cap < 1 {
		cap = 1
	}
	p := &Pool{
		queue: make(chan int, cap),
		wg:    new(sync.WaitGroup),
	}
	p.wg.Add(total)
	return p
}

func (p *Pool) AddOne() {
	p.queue <- 1
}

func (p *Pool) DelOne() {
	<-p.queue
	p.wg.Done()
}

// => pool block end }}}
var Usage = func(code int, msg ...string) {
	if len(msg) > 0 {
		fmt.Fprintf(os.Stdin, "\x1b[38;5;1m* %s\x1b[0m\n", msg[0])
	}
	fmt.Fprintf(os.Stdin, "Usage: %s [options] file/dir\n\n", filepath.Base(os.Args[0]))
	fmt.Println(" -e      encrypt file")
	fmt.Println(" -d      decrypt file")
	fmt.Println(" -w      worker count, default: 50")
	fmt.Println(" -x      dont encode filename")
	fmt.Println(" -h      show command usage")
	os.Exit(code)
}

func main() {
	ts := time.Now()

	cap := 50
	path := GetPathFromArgs(os.Args[1:])
	action := "encrypt"
	encfilename := true

	if ArraySearch(os.Args, "-h") > -1 {
		Usage(0)
	}

	if path == "" {
		Usage(2, "path is not exists, please check 'file/dir' param")
	}

	if ArraySearch(os.Args, "-d") > -1 {
		action = "decrypt"
	}

	if ArraySearch(os.Args, "-x") > -1 {
		encfilename = false
	}

	if wc := GetOptVal("-w"); wc != "" {
		nwc, _ := strconv.Atoi(wc)
		if nwc > 0 {
			cap = nwc
		}
	}

	filelist := GetFilelist(path)
	pool := NewPool(cap, len(filelist))
	for _, v := range filelist {
		go func(file string) {
			pool.AddOne()
			err := Encrypt(file, action, encfilename)
			if nil != err {
				P(err)
			}
			pool.DelOne()
		}(v)
	}
	pool.wg.Wait()

	te := time.Now()
	fmt.Printf("release time is: %.2f\n", te.Sub(ts).Seconds())
}
